working/macros/api/builders.dart

```diff
@@ -0,0 +1,115 @@
+import 'code.dart';
+import 'introspection.dart';
+import 'macros.dart'; // For dart docs :(
+
+abstract class Builder {
+  /// Used to construct a [TypeAnnotation] to a runtime type available to the
+  /// the macro implementation code.
+  ///
+  /// This can be used to emit a reference to it in generated code, or do
+  /// subtype checks (depending on support in the current phase).
+  TypeAnnotation typeAnnotationOf<T>();
+}
+
+/// The api used by [TypeMacro]s to contribute new type declarations to the
+/// current library, and get [TypeAnnotation]s from runtime [Type] objects.
+abstract class TypeBuilder implements Builder {
+  /// Adds a new type declaration to the surrounding library.
+  void addTypeToLibary(DeclarationCode typeDeclaration);
```

Comment:
Bikeshed: Maybe `declareType()`?

Reply:
Done, I also renamed some other things to follow that naming a bit more, `addToLibrary => declareInLibrary` and `addToClass => declareInClass`.